.buildkite/generate-pipeline.sh

```diff
@@ -0,0 +1,21 @@
+#! /usr/bin/env nix-shell
+#! nix-shell shell.nix -i bash
+
+# SPDX-FileCopyrightText: 2021 TQ Tezos <https://tqtezos.com/>
+#
+# SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+set -euo pipefail
+
+git fetch origin master
+
+diff="$(git diff origin/master...HEAD --name-only)"
+
+# Diff with the master branch is empty when we're in the master branch,
+# so we're using 'git diff --name-only HEAD HEAD~1' instead
+if [[ -z $diff ]]; then
+  diff="$(git diff --name-only HEAD HEAD~1)"
```

Comment:
Wouldn't this mean that the run on `master` would only be on the last commit's changes tho?

E.g. if I merge a PR with multiple commits where the last one is only a doc update, would the `master` pipeline not run any build?

Reply:
Somewhat related: this solution is based on:
> Pipeline from the yaml file is loaded to the python dictionary, this dict is later filtered based on the output of 'git diff

What if instead we'd let buildkite do the filtering of the steps with `if`s and we'd only fill the conditions from python?

E.g. in our `pipeline.yml` we have a step:
```yaml
 - label: crossref-verify
   command: [...]
   soft_fail: true
```
say that we want to run it on `master` in any case and on PRs only on doc changes, we could modify it to:
we can change it to:
```yaml
 - label: crossref-verify
   if: build.branch == "master" || DOC_CHANGE
   command: [...]
   soft_fail: true
```
and from our python code we can then do:
- if diff contains doc changes, replace the `DOC_CHANGE` string with `true`
- otherwise replace it with `false`

It seems to me it would be much easier to maintain as well, what do you think?

Reply:
Also, I think such a solution could also be used to decide weather or not to build native packages, as we were discussing previously.

We could have a variable in the python code that decides to change the "variables" in the yaml to `true` or `false` and so change that instead of (un)commenting lines in `pipeline.yml`.

Reply:
> Wouldn't this mean that the run on master would only be on the last commit's changes tho?

No, AFAIU, it will show the diff for the merge commit, which is exactly the contents of the PR that merged last

Reply:
> What if instead we'd let buildkite do the filtering of the steps with ifs and we'd only fill the conditions from python?

That should work, but doesn't sound very portable, i.e. it will for for this specific project. Additionally, python script may have to be adjusted each time we update something in the repository structure. IMO, updating `pipeline.yml` is a bit more convenient

Reply:
> it will show the diff for the merge commit

Yep, it seems you are right, I didn't check that properly :ok_hand: 

> doesn't sound very flexible
> updating pipeline.yml is a bit more convenient

Ok, that's a good point, however I still think there would be advantages in letting buidkite do the filtering with its own options.

For example, I see that we replicate the `only_changes` for dependent steps and that you mentioned:
> if it worth managing transitive only_changes dependencies automatically instead

which we could say happens because the filtering goes on before the pipeline gets "evaluated" by buildkite.

So, I have a counter-proposal, considering that from [the buildkite docs](https://buildkite.com/docs/pipelines/dependencies#order-of-operations):
> However, if the step you're dependent on is excluded from the build due to an `if` condition, the dependency will be ignored and the step that depends on it will run once any other dependencies are satisfied.

what if instead of filtering the steps in the python script we use the `only_changes` to edit/create the `if` attribute?

So we run the regex(es) from the `only_changes`, if the result is `false` we add/replace the `if` of that same step with `"false" # skipped due to lack of changes`, otherwise we leave it alone (and either way we remove the `only_changes`).

This way we still get to update the pipeline only by editing the `yml`, but we also get the steps to not care about their dependencies `only_changes`.

Reply:
> what if instead of filtering the steps in the python script we use the only_changes to edit/create the if attribute?

There is a `skip` attribute that will probably do what we want instead of potentially extending existing `if`

> the filtering goes on before the pipeline gets "evaluated" by buildkite.

Yeah, that's a good point why its worth using `if` instead of filtering pipeline manually

Reply:
> what if instead of filtering the steps in the python script we use the only_changes to edit/create the if attribute?

I've just realized that there is a problem with the reverse transitive dependencies, that buildkite won't be able to solve with the `if`. Imagine the following: something in `tests` was updated, thus `build-via-docker` won't be triggered along with `test docker-built binaries`, which is supposed to be triggered due to changes in `tests`

Reply:
One more thing: `if` cannot be used together with `branches` attribute, thus all `branches` occurrences need to be reimplemented via `if`

Reply:
> something in tests was updated, thus build-via-docker won't be triggered along with test docker-built binaries, which is supposed to be triggered due to changes in tests.

:thinking: it seems to me that would be correctly handled tho.

`build-via-docker` won't be executed as you said because of the `if`, which I'd assume is ok as we don't want to rebuild it, only run new test. Then for the point I linked above:
>  if the step you're dependent on is excluded from the build due to an if condition, the dependency will be ignored and the step that depends on it will run

The `test docker-built binaries` should be executed, as its `if` was not modified by the script.

However I see now that the latter depends on the artifact from the former, so perhaps that's the issue?

> also vice-versa: build-via-docker was triggered because tezos revision was updated, but test docker-built binaries weren't because there are no changes in tests

Looks like this is a case where `only_changes` should include transient dependencies, but not just because the step its depending on needs them, but logically because if the source has changed then tests needs to be run as well.

P.s. from what I can tell these aren't issue of `if` only, but would appear with the pre-filtering as well.

Reply:
> from what I can tell these aren't issue of if only, but would appear with the pre-filtering as well.

This issue is indeed present in the pre-filtering case and currently, it's solved by listing transitive dependencies explicitely